From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Legitimoose <legitimoose@gmail.com>
Date: Sun, 23 Jun 2024 15:24:39 -0700
Subject: [PATCH] Make /execute only run vanilla commands

Adds a new command dispatcher and root that clones the vanilla, non-blacklisted commands.
/execute run can only use this root.

diff --git a/src/main/java/net/minecraft/commands/Commands.java b/src/main/java/net/minecraft/commands/Commands.java
index 476914d5448799c5174956e0b8b4b534af235862..c04d8867d340a8e353117f36784b2341577df006 100644
--- a/src/main/java/net/minecraft/commands/Commands.java
+++ b/src/main/java/net/minecraft/commands/Commands.java
@@ -1,6 +1,7 @@
 package net.minecraft.commands;
 
 import com.google.common.collect.Maps;
+import com.mojang.brigadier.CommandDispatcher; // Moose
 import com.legitimoose.lsp.MooseConfig; // Moose
 import com.mojang.brigadier.ParseResults;
 import com.mojang.brigadier.StringReader;
@@ -156,6 +157,7 @@ public class Commands {
     public static final int LEVEL_ADMINS = 3;
     public static final int LEVEL_OWNERS = 4;
     private final com.mojang.brigadier.CommandDispatcher<CommandSourceStack> dispatcher = new com.mojang.brigadier.CommandDispatcher();
+    public static CommandDispatcher<CommandSourceStack> vanillaCommands = new CommandDispatcher<>(); // Moose
 
     public Commands(Commands.CommandSelection environment, CommandBuildContext commandRegistryAccess) {
         // Paper
@@ -271,6 +273,15 @@ public class Commands {
 //            this.vanillaCommandNodes.add(node);
         }
         // Moose end
+        // Moose start - Copy all the vanilla commands into a new root for use in /execute. Prevents plugin commands being run.
+        vanillaCommands.getRoot().clearAll(); // Moose - Allow reloading of RessourceOrId arguments (like in /loot command)
+        for (CommandNode<CommandSourceStack> node : this.dispatcher.getRoot().getChildren()) {
+            if (!node.commandBlockAllowed) {
+                continue;
+            }
+            vanillaCommands.getRoot().addChild(node);
+        }
+        // Moose end
 
         // Paper start - Vanilla command permission fixes
         for (final CommandNode<CommandSourceStack> node : this.dispatcher.getRoot().getChildren()) {
diff --git a/src/main/java/net/minecraft/server/commands/ExecuteCommand.java b/src/main/java/net/minecraft/server/commands/ExecuteCommand.java
index b4c21996550da123d9e0f46e0a7ef67eb8f4ff76..df3dca13a01c25762a09836a481c35288a0dd8ff 100644
--- a/src/main/java/net/minecraft/server/commands/ExecuteCommand.java
+++ b/src/main/java/net/minecraft/server/commands/ExecuteCommand.java
@@ -144,7 +144,7 @@ public class ExecuteCommand {
         dispatcher.register(
             Commands.literal("execute")
                 .requires(source -> source.hasPermission(2))
-                .then(Commands.literal("run").redirect(dispatcher.getRoot()))
+                .then(Commands.literal("run").redirect(Commands.vanillaCommands.getRoot())) // Moose - Only run vanilla commands
                 .then(addConditionals(literalCommandNode, Commands.literal("if"), true, commandRegistryAccess))
                 .then(addConditionals(literalCommandNode, Commands.literal("unless"), false, commandRegistryAccess))
                 .then(Commands.literal("as").then(Commands.argument("targets", EntityArgument.entities()).fork(literalCommandNode, context -> {
